### PR TITLE
under the hood edit

### PR DESCRIPTION
**Pull request by zmzlois**
with two to-dos

### CODE REVIEW

File: docs/advanced/under-the-hood.md [hunk 1]
```diff
@@ -2,10 +2,19 @@
...
+
+Although we are extremely hyped to talk about what we created, we find this article will be much more understandable if you have built something with ZenStack, following a simple guide like [Todo SaaS Example](https://zenstack.dev/docs/get-started/full-todo-app), and come to check this article with questions, otherwise some ideas like Auxiliary Field might be a bit...let's say, densed.
 
 # How Does ZenStack Work Under the Hood?
 
-ZenStack extends Prisma ORM mainly at two levels. First, it provides a modeling DSL (ZModel) - a superset of Prisma schema. Second, it allows the creation of enhanced wrappers for Prisma clients at runtime for injecting extra behaviors.
+ZenStack extends Prisma ORM mainly at two levels through our creation of ZModel, a modelling [DSL](https://en.wikipedia.org/wiki/Domain-specific_language) from scratch. It is firstly a superset of Prisma and cover the core functionality of Prisma.
```
Comment (ymc9) [inline docs/advanced/under-the-hood.md:11]: The point I tried to make is that ZenStack extends Prisma at compile time (through ZModel) and at runtime (through enhancements). I'll see if I can split it into two paragraphs to make it easier to understand.

File: docs/advanced/under-the-hood.md [hunk 2]
```diff
@@ -25,9 +36,9 @@ ZenStack implemented the ZModel DSL from scratch, including the CLI and the VSCo
 
 -   Policy: `@core/policy`
 
-    The policy plugin converts access policy rules (expressed with `@@allow` and `@@deny` attributes) into checker functions. The functions take a context object as input and return partial Prisma query objects, which will be injected into Prisma query arguments at runtime. The context object contains the following properties:
+The policy plugin converts access policy rules (expressed with `@@allow` and `@@deny` attributes) into checker functions (the function checks whether a user is logged in, or has certain permissions to access a specific part of the app or data).. The functions take a context object as input and return partial Prisma query objects, which will be injected into Prisma query arguments at runtime. The context object contains the following properties:
```
Comment (ymc9) [inline docs/advanced/under-the-hood.md:39]: I'll add an example here.

File: docs/advanced/under-the-hood.md [hunk 3]
```diff
@@ -315,3 +358,6 @@ When ZModel is transpiled to Prisma schema, two auxiliary fields are added to ea
     To support such cases, we use a transaction to wrap the mutation and set the `zenstack_transaction` field to a unique value containing a CUID and the operation (e.g., "create:clg4szzhq000008jf6ppybhb6"). Then, when checking rules, we can query the database (inside the transaction) to find out which entities are affected by the mutation.
 
 The auxiliary fields add some storage and computation overheads to the database, and we can consider optimizing their usage or even removing them in the future.
+
+---
+Again if you find the later part of this article hard to follow, we STRONGLY RECOMMEND YOU go through our examples, build a simple app, hit `npx zenstack generate` and read the code comes out -- learning with questions is always faster.
```
Comment (ymc9) [inline docs/advanced/under-the-hood.md:363]: I think I can merge this PR to a feature branch and continue making some tweaks there before publishing it. Thanks for work on this @zmzlois !

**Comment by ymc9** [comment]
Much of the explanations have been contained in the new "The Complete Guide" docs. Closing this PR for now. Thanks for making this effort!

**Comment by zmzlois** [comment]
You are welcome! Would love to contribute if you feel like anything can be improved but not sure about which part can be better!


Sent via ***@***.***>


On Sat, Dec 02, 2023 at 04:24:19, Yiming ***@***.******@***.***>> wrote:

Much of the explanations have been contained in the new "The Complete Guide" docs. Closing this PR for now. Thanks for making this effort!
